packages/dds/tree/src/events/events.ts

```diff
@@ -252,26 +255,25 @@ export class EventEmitter<E extends Events<E>> implements ISubscribable<E>, HasL
 	 * invoking the returned callback can error even if its only called once if the same listener was provided to two calls to "on".
 	 * This behavior is not documented and its unclear if its a bug or not: see note on listeners.
 	 */
-	public on<K extends keyof Events<E>>(eventName: K, listener: E[K]): () => void {
-		getOrCreate(this.listeners, eventName, () => new Set()).add(listener);
-		return () => this.off(eventName, listener);
-	}
-
-	private off<K extends keyof Events<E>>(eventName: K, listener: E[K]): void {
-		const listeners =
-			this.listeners.get(eventName) ??
-			// TODO: consider making this (and assert below) a usage error since it can be triggered by users of the public API: maybe separate those use cases somehow?
-			fail(
-				"Event has no listeners. Event deregistration functions may only be invoked once.",
+	public on<K extends keyof Events<E>>(eventName: K, listener: E[K]): Off {
+		const off: Off = () => {
+			const listeners =
+				this.listeners.get(eventName) ??
+				// TODO: consider making this (and assert below) a usage error since it can be triggered by users of the public API: maybe separate those use cases somehow?
+				fail(
+					"Event has no listeners. Event deregistration functions may only be invoked once.",
+				);
+			assert(
+				listeners.delete(off),
+				0x4c1 /* Listener does not exist. Event deregistration functions may only be invoked once. */,
```

Comment:
The fail and the assert feel inverted to me. If we don't have an array of listeners for `eventName` when `off` executes, something must have gone really wrong...? But if someone called `off` twice, the first time it was removed from the aray of listeners, then trying to remove it again feels like it _could_ be a no-op instead of an assertion failure.

Reply:
I totally agree. We discussed this at API council recently, and I'm convinced it should be a no-op. The policy above is not new - it's moved/copied from the existing implementation - and I avoided changing it in this PR to keep the PR more pure. But I'll be doing a followup PR that will rectify that. I have a little list of event things to improve, this is just the first :)